django/db/models/sql/query.py

```diff
@@ -1299,54 +1308,62 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
             self.check_filterable(value)
 
         clause = self.where_class()
+        alias = self.get_initial_alias()
+
         if reffed_expression:
             condition = self.build_lookup(lookups, reffed_expression, value)
             clause.add(condition, AND)
-            return clause, []
-
-        opts = self.get_meta()
-        alias = self.get_initial_alias()
-        allow_many = not branch_negated or not split_subq
+            # When col is nullable, add IS NOT NULL.
+            col = get_col_from_expression(reffed_expression)
+            if col:
+                lookup_type = condition.lookup_name
+                target = col.target
+            else:
+                return clause, ()
+        else:
```

Comment:
I really don't like that we increase indentation here, it make code less readable, and it's complicated even without this :disappointed:  We could reduce the number of changes significantly with:
```python
        if reffed_expression:
            condition = self.build_lookup(lookups, reffed_expression, value)
            clause.add(condition, AND)
            # When col is nullable, add IS NOT NULL.
            col = self._gen_cols(reffed_expression)
            if col:
                lookup_type = condition.lookup_name
                target = col.target
                if current_negated and (lookup_type != 'isnull' or condition.rhs is False) and condition.rhs is not None and self.is_nullable(target):
                    lookup_class = target.get_lookup('isnull')
                    col = self._get_col(target, target, alias)
                    clause.add(lookup_class(col, False), AND)
            return clause, ()
```
and reverting related adjustments.

Reply:
@felixxm Thank you for reviewing. we also have to handle when col is 'Subquery'.

Reply:
Can we add a test for this case? Currently all tests work with [my proposition](https://github.com/django/django/pull/14065#discussion_r635015330).

Reply:
@felixxm 
It doesn't work when col is the subquery.
below case is failed.

```
    def test_exclude_nullable_subquery_annotation(self):
        number = Number.objects.create(num=1, other_num=1)
        subquery_with_null = Subquery(Number.objects.filter(num=1).values('another_num')[:1])
        self.assertSequenceEqual(
            Number.objects.annotate(current_num=subquery_with_null).exclude(current_num=1),
            [number]
        )
```
Second sequence contains 1 additional elements.
First extra element 0:
<Number: 1>

- <QuerySet []>
+ [<Number: 1>]

----------------------------------------------------------------------
Ran 1 test in 0.005s

Reply:
I updated the code according to your suggestion. my test case(for ticket-32864) doesn't work. if there is any missing point, please tell me. and many errors occurred with  "target = col.target
AttributeError: 'generator' object has no attribute 'target'".

Reply:
I tried to handle when reffed_expression is the subquery. but so many tests are failed.

Reply:
FYI I ran into the same issue of `AttributeError: 'generator' object has no attribute 'target'".` with this test checking if this PR fixed the issue when using exclude with an `alias` (flagged up in duplicate [ticket-32896](https://code.djangoproject.com/ticket/32896)):
```python
    # ExcludeTests

    def test_exclude_aliased_nullable_fields(self):
        number = Number.objects.create(num=1, other_num=1)
        Number.objects.create(num=2, other_num=2, another_num=2)
        qs = Number.objects.alias(aliased_num=F('num'))
        self.assertSequenceEqual(
            qs.exclude(another_num=F('aliased_num')),
            [number],
        )
        self.assertSequenceEqual(
            qs.exclude(aliased_num=F('another_num')),
            [number],
        )
```